libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp

```diff
@@ -471,6 +490,179 @@ uint64_t VoteManager::roundDeterminedFromVotes(size_t two_t_plus_one) {
   return 0;
 }
 
+void VoteManager::sendRewardPeriodCertVotes(uint64_t reward_period) {
+  auto reward_period_cert_votes = db_->getCertVotes(reward_period);
+  auto net = network_.lock();
+  net->onNewPbftVotes(std::move(reward_period_cert_votes), true);
+}
+
+bool VoteManager::AddRewardVote(const std::shared_ptr<Vote>& vote) {
+  // Don't check the vote if in DB here. That will check in updateRewardVotes for reducing DB read
+  const auto vote_hash = vote->getHash();
+  std::unique_lock lock(reward_votes_mutex_);
+  if (reward_votes_.second.contains(vote_hash)) {
+    return false;
+  }
+
+  const auto voted_block_hash = vote->getBlockHash();
+  const auto pbft_chain_last_block_hash = pbft_chain_->getLastPbftBlockHash();
+  if (voted_block_hash != pbft_chain_last_block_hash) {
+    LOG(log_dg_) << "Drop reward vote " << vote_hash << ". PBFT chain last block is " << pbft_chain_last_block_hash
+                 << ", but vote for PBFT block is " << voted_block_hash;
+    return false;
+  }
+
+  if (!reward_votes_.first) {
+    reward_votes_.first = voted_block_hash;
+  } else if (reward_votes_.first && reward_votes_.first != pbft_chain_last_block_hash) {
+    // Clear missing reward votes
+    LOG(log_dg_) << "Remove reward votes: \n";
+    for (const auto& v : reward_votes_.second) {
+      LOG(log_dg_) << v.first;
+    }
+    reward_votes_ = {};
+    reward_votes_.first = voted_block_hash;
+  }
+  assert(voted_block_hash == reward_votes_.first);
+
+  reward_votes_.second.insert({vote_hash, vote});
+  LOG(log_nf_) << "add reward vote " << vote_hash;
+
+  return true;
+}
+
+bool VoteManager::verifyRewardVote(std::shared_ptr<Vote>& vote) {
+  const auto voted_block_hash = vote->getBlockHash();
+  const auto pbft_chain_last_block_hash = pbft_chain_->getLastPbftBlockHash();
+  if (voted_block_hash != pbft_chain_last_block_hash) {
+    LOG(log_er_) << "Drop reward vote " << vote->getHash() << ". PBFT chain last block is "
+                 << pbft_chain_last_block_hash << ", but vote for PBFT block is " << voted_block_hash;
+    return false;
+  }
+
+  if (vote->getType() != cert_vote_type) {
+    LOG(log_er_) << "Reward vote type is not cert vote type " << *vote;
+    return false;
+  }
+
+  if (vote->getStep() != 3) {
```

Comment:
is this correct? maybe I am wrong but step can not be higher?

Reply:
cert vote step is only 3, 4+ steps are next votes